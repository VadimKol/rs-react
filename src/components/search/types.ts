export interface SearchProps {
  searchField: React.RefObject<HTMLInputElement>;
  character: { name: string };
  setCharacter: (character: { name: string }) => void;
  loader: boolean;
}
