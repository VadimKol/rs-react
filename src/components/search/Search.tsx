import { type ReactNode } from 'react';
import { useSearchParams } from 'react-router-dom';

import { useLocalStorage } from '@/hooks/useLocalStorage';

import styles from './styles.module.scss';
import type { SearchProps } from './types';

export function Search({ searchField, character, setCharacter, setPage, setLoader, loader }: SearchProps): ReactNode {
  const [, setPageQuery] = useSearchParams();
  const [, setLs] = useLocalStorage('R&M_search');
  return (
    <form
      className={styles.search_form}
      onSubmit={(e) => {
        e.preventDefault();
        if (typeof searchField.current?.value === 'string' && !loader) {
          const searchValue = searchField.current?.value.trim();
          setCharacter({ name: searchValue });
          setLs(searchValue);
          setPage(1);
          setLoader(true);
          setPageQuery({ page: '1' });
        }
      }}
    >
      <input
        id="search"
        className={styles.search}
        type="text"
        placeholder="Search..."
        ref={searchField}
        defaultValue={character.name}
      />
      <button data-testid="search" type="submit" className={styles.search_button} aria-label="Search-button" />
    </form>
  );
}
