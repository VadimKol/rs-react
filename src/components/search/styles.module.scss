.search_form {
  display: flex;
  justify-content: center;

  & .search_box {
    position: relative;
  }

  & .search {
    box-sizing: border-box;
    width: 400px;
    height: 45px;
    padding: 0 40px 0 5px;
    font-family: var(--primary-font);
    font-size: 30px;
    color: var(--text-color);
    background: var(--background-color) padding-box;
    border: 2.5px solid transparent;
    border-radius: 6px;

    &:focus {
      border: 2.5px solid var(--primary-color);
    }

    &:focus-visible {
      outline: none;
    }
  }

  & .search_button {
    position: absolute;
    top: 7px;
    right: 7px;
    width: 30px;
    height: 30px;
    padding: 0;
    background-color: var(--text-color);
    border: none;
    transition: background-color 0.5s;
    mask-image: url("/src/assets/images/search.svg");
    mask-size: cover;

    &:focus-visible {
      border-radius: 6px;
      outline: 2px solid var(--primary-color);
    }

    @media (hover: hover) {
      &:hover {
        background-color: var(--primary-color);
        cursor: pointer;
      }
    }
  }
}
