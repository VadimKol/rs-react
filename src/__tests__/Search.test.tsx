import '@testing-library/jest-dom';

import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { type RefObject } from 'react';
import { MemoryRouter } from 'react-router-dom';

import { Search } from '@/components/search/Search';

const searchField: RefObject<HTMLInputElement> = { current: document.createElement('input') };

if (searchField.current) {
  searchField.current.value = 'Rick';
}

describe('Search Component', () => {
  const user = userEvent.setup();

  afterEach(() => {
    localStorage.removeItem('R&M_search');
  });

  it('renders correctly', () => {
    const { container } = render(
      <MemoryRouter>
        <Search
          searchField={searchField}
          character={{ name: 'Rick' }}
          setCharacter={jest.fn}
          setPage={jest.fn}
          setLoader={jest.fn}
          loader={false}
        />
      </MemoryRouter>,
    );

    expect(container).toMatchSnapshot();
  });

  it('checks searchTerm in localStorage', async () => {
    render(
      <MemoryRouter>
        <Search
          searchField={searchField}
          character={{ name: 'Rick' }}
          setCharacter={jest.fn}
          setPage={jest.fn}
          setLoader={jest.fn}
          loader={false}
        />
      </MemoryRouter>,
    );

    const buttonSearch = screen.getByRole('button', { name: 'Search-button' });

    await user.click(buttonSearch);

    expect('Rick').toBe(localStorage.getItem('R&M_search'));
  });
});
