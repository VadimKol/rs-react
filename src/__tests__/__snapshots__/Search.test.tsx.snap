// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Component renders correctly 1`] = `
<div>
  <form
    class="search_form"
  >
    <input
      class="search"
      id="search"
      placeholder="Search..."
      type="text"
      value="Rick"
    />
    <button
      aria-label="Search-button"
      class="search_button"
      data-testid="search"
      type="submit"
    />
  </form>
</div>
`;
